src/async/bimap/bimap.js

```diff
@@ -5,14 +5,14 @@
  * @param leftFn {Function}
  * @param rightFn {Function}
  * @param p {Promise<T | never>}
- * @returns {Promise<T | never>}
+ * @returns {Promise<R | never>}
  */
-function bimap<T>(
-  leftFn: (Error) => any,
-  rightFn: (T) => any,
+function bimap<T, R, E>(
+  leftFn: (E) => any,
+  rightFn: (T) => R,
   p: Promise<T>
-): Promise<T> {
-  return p.then(rightFn, leftFn);
+): Promise<R> {
+  return p.then(rightFn, (e: E): Promise<R> => Promise.reject(leftFn(e)));
```

Comment:
I wasn't sure what to do here.  I still half-feel this is correct.

If we were to write out the `rightFn` it would look like this

function bimap<T, R, E>(
  leftFn: (E) => any,
  rightFn: (T) => R,
  p: Promise<T>
): Promise<R> {
  return p.then(
    **(v: T): Promise<R> => Promise.resolve(rightFn(v)),**
    (e: E): Promise<R> => Promise.reject(leftFn(e)),
  );
}

Since rightFn returns Promise<R>, shouldn't leftFn return Promise<R> too – so that the function on both sides returns the same type?

(If leftFn returned Promise<any>, would bimap then return Promise<any>?)

Reply:
Oh I see the confusion. So promises only annotate the Success value, not the error.  So as long as the promise on the left is rejected we cannot type it.
https://github.com/facebook/flow/issues/1232
https://github.com/facebook/flow/issues/5042

Alternatively we could say, incorrectly, Promise<R | SomeErrorType> but that's just us being super clever :)

Reply:
Yeah that's my understanding.  Which is a huge caveat of this approach, but I still think there's a lot of benefits to it which is why I'm PRing :).

Unfortunately that union trick won't work for long:

https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AVQM5YBOoAvKAN6gB2AhgLZYBco+KRAltQOagC+AblSYcAUSJFYJclQDGsACbMaAV3oAjYvyGJIK6rJQdY1UNywoCxAEJoAkgoAUHBQEoWABUn0OhADxWJAA+oOKSRAB8lKBEFipEpl6wPoQAdLH4sNAAbliOVHSMLADkAMpYtNTF-K4C-Mjy1GygKoREnt6+WAFtoCFhUlHk5pZttg6OAIy1yK3EqSgAFljUjo5LWIEsga5kURSIoKCNmdBYqXDc68uBqYVYM3yuqbK0KLKLa8TtoRJSu6R9odjiZTudLo5ivIVNAFDRYCgzBYWm1qgBqUDfF6KB5CJ5AA

As a comment to the issues around being unable to type the rejection because anything might be thrown, as long as every Promise has a `catch((err: any) => ConcreteErrVal)` defined, shouldn't it then be possible to type the resulting Promises' rejected side with that `Promise<SuccessType, ConcreteErrVal>`.  (And prior to that catch, it would have to be `Promise<SuccessVal, any>`?)

Getting off topic... how should we resolve the bimap?